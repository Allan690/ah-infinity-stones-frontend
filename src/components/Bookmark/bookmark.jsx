import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bookmarkRequest } from '../../redux/actions/bookmarkActions';
import './bookmarks.scss';
import { successMessage, errorMessage } from '../../helpers/messages';

export class BookMark extends PureComponent {
  state = { className: 'fas fa-bookmark' };

  static propTypes = {
    bookmarkRequest: PropTypes.func.isRequired,
    slug: PropTypes.string.isRequired,
    message: PropTypes.string,
    error: PropTypes.string,
    prevMessage: PropTypes.string,
    prevError: PropTypes.string,
  };

  static defaultProps = {
    message: '',
    error: '',
    prevMessage: '',
    prevError: '',
  };

  componentDidUpdate(prevProps) {
    const { message, error } = this.props;
    const { prevMessage, prevError } = prevProps;
    if (message !== prevMessage && message !== null) {
      this.handleSuccess(message);
    } else if (prevError !== error) {
      this.handleError(error);
    }
  }

  handleBookMark = (event) => {
    const {
      slug, bookmarkRequest: bookmark,
    } = this.props;
    event.preventDefault();
    bookmark(slug);
  };

  handleSuccess = (message) => {
    successMessage(message);
    this.setState({ className: 'fas fa-bookmark' });
  };

  handleError = (error) => {
    errorMessage(error);
    this.setState({ className: 'far fa-bookmark' });
  };

  render() {
    const { className } = this.state;
    return (
      <div>
        <button
          type='button'
          className='fav-btn'
          onClick={this.handleBookMark}
        >
          <i className={className} />
          &nbsp;
          <strong>Bookmark</strong>
        </button>

      </div>
    );
  }
}

const mapDispatchToProps = { bookmarkRequest };

const mapStateToProps = (state) => {
  const { bookmarkReducer: { message, error } } = state;

  return { message, error };
};


export default connect(mapStateToProps, mapDispatchToProps)(BookMark);
