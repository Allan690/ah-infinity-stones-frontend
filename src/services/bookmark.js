import {
  baseURL,
} from '../common/constants';

export async function bookmark(slug) {
  try {
    const response = await fetch(`${baseURL}articles/${slug}/bookmark`, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-type': 'application/json',
        Authorization: `${localStorage.getItem('Token')}`,
      },
    });
    const data = await response.json();
    return data;
  } catch (error) {
    return error;
  }
}

export async function unbookmark(slug) {
  try {
    const response2 = await fetch(`${baseURL}articles/${slug}/bookmark`, {
      method: 'DELETE',
      headers: {
        Accept: 'application/json',
        'Content-type': 'application/json',
        Authorization: `${localStorage.getItem('Token')}`,
      },
    });
    const data = await response2.json();
    return data;
  } catch (error) {
    return error;
  }
}
