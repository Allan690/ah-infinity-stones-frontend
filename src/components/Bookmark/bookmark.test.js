import React from 'react';
import { BookMark } from './bookmark';


describe('Bookmark Component', () => {
  const preventDefault = jest.fn();
  const props = {
    bookmarkRequest: jest.fn(),
  };

  const wrapper = shallow(<BookMark {...props} />);

  it('Should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should find bookmark button', () => {
    expect(wrapper.find('.fav-btn').exists()).toBeTruthy();
  });

  it('Should find bookmark button', () => {
    wrapper.find('.fav-btn').simulate('click', { preventDefault });
    expect(props.bookmarkRequest).toHaveBeenCalled();
  });
});
