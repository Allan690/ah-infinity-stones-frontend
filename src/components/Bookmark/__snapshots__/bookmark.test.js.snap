// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark Component Should render properly 1`] = `
<div>
  <button
    className="fav-btn"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fas fa-bookmark"
    />
    Â 
    <strong>
      Bookmark
    </strong>
  </button>
</div>
`;
